File: fairdiplomacy/data/data_fields.py
# Copyright (c) Facebook, Inc. and its affiliates.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.

import torch
from fairdiplomacy.utils.tensorlist import TensorList


class DataFields(dict):
    BOOL_STORAGE_FIELDS = ["x_board_state", "x_prev_state"]

    def select(self, idx):
        return DataFields({k: v[idx] for k, v in self.items()})

    @classmethod
    def cat(cls, L: list):
        if len(L) > 0:
            return cls({k: _cat([x[k] for x in L]) for k in L[0]})
        else:
            return cls()

    @classmethod
    def stack(cls, L: list, dim: int = 0):
        if len(L) > 0:
            return cls({k: torch.stack([x[k] for x in L], dim) for k in L[0]})
        else:
            return cls()

    def to_storage_fmt_(self):
        for f in DataFields.BOOL_STORAGE_FIELDS:
            self[f] = self[f].to(torch.bool)
        return self

    def from_storage_fmt_(self):
        for f in DataFields.BOOL_STORAGE_FIELDS:
            self[f] = self[f].to(torch.float32)
        return self

    def to(self, *args, **kwargs):
        return DataFields(
            {k: v.to(*args, **kwargs) if hasattr(v, "to") else v for k, v in self.items()}
        )


def _cat(x):
    return TensorList.cat(x) if isinstance(x[0], TensorList) else torch.cat(x)

File: dipcc/dipcc/cc/data_fields.cc
/*
Copyright (c) Facebook, Inc. and its affiliates.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
*/

#include "data_fields.h"

namespace dipcc {

TensorDict new_data_fields_state_only(long B) {
  return {
      {"x_board_state", torch::empty({B, 81, 35}, torch::kFloat32)},
      {"x_prev_state", torch::empty({B, 81, 35}, torch::kFloat32)},
      {"x_prev_orders", torch::empty({B, 2, 100}, torch::kLong)},
      {"x_season", torch::empty({B, 3}, torch::kFloat32)},
      {"x_in_adj_phase", torch::empty({B}, torch::kFloat32)},
      {"x_build_numbers", torch::empty({B, 7}, torch::kFloat32)},
  };
}

TensorDict new_data_fields(long B, long max_seq_len, bool include_power) {
  TensorDict fields(new_data_fields_state_only(B));
  fields["x_loc_idxs"] = torch::full({B, 7, 81}, -1, torch::kInt8);
  fields["x_possible_actions"] =
      torch::full({B, 7, max_seq_len, 469}, -1, torch::kInt32);

  if (include_power) {
    fields["x_power"] = torch::full({B, 7, max_seq_len}, -1, torch::kLong);
  }

  return fields;
}

} // namespace dipcc
